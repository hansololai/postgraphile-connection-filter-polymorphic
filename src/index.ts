export * from './pgDefinePolymorphicCustomPlugin';
export * from './pgDefineTableToModelMapPlugin';
export * from './pgConnectionArgFilterForwardPolyRelationPlugin';
export * from './pgConnectionArgFilterBackwardPolyRelationPlugin';
