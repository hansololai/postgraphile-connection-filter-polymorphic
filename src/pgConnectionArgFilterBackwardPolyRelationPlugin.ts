import { SchemaBuilder, Options } from 'postgraphile';
import {
  PgPolymorphicConstraintByName, PgPolymorphicConstraint,
} from './pgDefinePolymorphicCustomPlugin';
export interface BackwardPolyRelationSpecType {
  table: any;
  foreignTable: any;
  fieldName: string;
  tablePrimaryKey: any;
  constraint: PgPolymorphicConstraint;
  isOneToMany: boolean;
}

const getSqlSelectWhereKeysMatch = ({ sourceAlias,
  foreignTableAlias, foreignTable, table, constraint, tablePrimaryKey, sql, inflection,
}) => {
  const sourceTableId = `${constraint.name}_id`;
  const sourceTableType = `${constraint.name}_type`;
  const tableTypeName = inflection.tableType(table);
  const sqlIdentifier = sql.identifier(
    foreignTable.namespace.name,
    foreignTable.name,
  );

  const sqlKeysMatch = sql.query`(${sql.fragment`${foreignTableAlias}.${sql.identifier(
    sourceTableId,
  )} = ${sourceAlias}.${sql.identifier(tablePrimaryKey.name)}`}) and (
  ${sql.fragment`${foreignTableAlias}.${sql.identifier(
    sourceTableType,
  )} = ${sql.value(tableTypeName)}`})`;
  const sqlSelectWhereKeysMatch = sql.query`select 1 from ${sqlIdentifier} as
  ${foreignTableAlias} where ${sqlKeysMatch}`;

  return sqlSelectWhereKeysMatch;
};
export const addField = (
  fieldName, description, type, resolve, spec, hint, build, fields,
  relationSpecByFieldName, context,
) => {
  const { extend, connectionFilterRegisterResolver } = build;
  const { fieldWithHooks, Self } = context;
  // Field
  const toReturn = extend(
    fields,
    {
      [fieldName]: fieldWithHooks(
        fieldName,
        {
          description,
          type,
        },
        {
          isPgConnectionFilterField: true,
        },
      ),
    },
    hint,
  );
  // Relation spec for use in resolver
  relationSpecByFieldName[fieldName] = spec;
  // Resolver
  connectionFilterRegisterResolver(Self.name, fieldName, resolve);
  return toReturn;
};
export const addBackwardPolyRelationFilter = (builder: SchemaBuilder, option: Options) => {
  // First add an inflector for polymorphic backrelation type name
  builder.hook('inflection', inflection => ({
    ...inflection,
    filterManyPolyType(table, foreignTable) {
      return `${this.filterManyType(table, foreignTable)}Poly`;
    },
    backwardRelationByPolymorphic(table, polymorphicName: string, isUnique: boolean) {
      const fieldName = isUnique ? this.singularize(table.name) : table.name;
      return this.camelCase(`${fieldName}-as-${polymorphicName}`);
    },
  }));
  const { pgSimpleCollections } = option;
  const hasConnections = pgSimpleCollections !== 'only';
  builder.hook('GraphQLInputObjectType:fields', (fields, build, context) => {
    const {
      describePgEntity,
      extend,
      newWithHooks,
      inflection,
      pgOmit: omit,
      pgSql: sql,
      pgIntrospectionResultsByKind: introspectionResultsByKind,
      graphql: { GraphQLInputObjectType, GraphQLBoolean },
      connectionFilterResolve,
      connectionFilterRegisterResolver,
      connectionFilterTypesByTypeName,
      connectionFilterType,
      mapFieldToPgTable,
      pgPolymorphicClassAndTargetModels = [],
    } = build;
    const {
      fieldWithHooks,
      scope: { pgIntrospection: table, isPgConnectionFilter },
      Self,
    } = context;

    if (!isPgConnectionFilter || table.kind !== 'class') return fields;
    if (!mapFieldToPgTable) {
      throw new Error(
        'mapFieldToPgTable is not defined in build, \
        you might missed plugin to run before this plugin',
      );
    }

    let newFields = fields;
    connectionFilterTypesByTypeName[Self.name] = Self;

    const modelName = inflection.tableType(table);

    const backwardRelationSpecs = (<PgPolymorphicConstraintByName>pgPolymorphicClassAndTargetModels)
      .filter(con => con.to.includes(modelName))
      // .filter((con) => con.type === 'f')
      // .filter((con) => con.foreignClassId === table.id)
      .reduce((memo, currentPoly) => {
        // if (omit(foreignConstraint, 'read')) {
        //   return memo;
        // }
        const foreignTable = introspectionResultsByKind.classById[currentPoly.from];
        if (!foreignTable) {
          return memo;
          // throw new Error(
          //   `Could not find the foreign table (polymorphicName: ${currentPoly.name})`,
          // );
        }
        if (omit(foreignTable, 'read')) {
          return memo;
        }
        const primaryConstraint = introspectionResultsByKind.constraint.find(
          attr => attr.classId === table.id && attr.type === 'p',
        );
        const sourceTableId = `${currentPoly.name}_id`;
        const sourceTableType = `${currentPoly.name}_type`;
        const isForeignKeyUnique = introspectionResultsByKind.constraint.find((c) => {
          // Only if the xxx_type, xxx_id are unique constraint
          // It must be an unique constraint
          if (c.classId !== foreignTable.id
            || c.keyAttributeNums.length !== 2
            || c.type !== 'u'
            || !c.keyAttributes.find(a => a.name === sourceTableId)
            || !c.keyAttributes.find(a => a.name === sourceTableType)
          ) return false;
          // the two attributes must be xx_type, xx_id
          return true;
        });
        const fieldName = inflection.backwardRelationByPolymorphic(
          foreignTable, currentPoly.name, isForeignKeyUnique,
        );
        // const fieldName = isForeignKeyUnique ? inflection.camelCase(
        //   inflection.singularize(foreignTable.name))
        //   : inflection.camelCase(inflection.pluralize(foreignTable.name));

        memo.push({
          table,
          fieldName,
          foreignTable,
          tablePrimaryKey: primaryConstraint.keyAttributes[0],
          isOneToMany: !isForeignKeyUnique,
          constraint: currentPoly,
        });
        return memo;
      }, [] as BackwardPolyRelationSpecType[]);

    const backwardRelationSpecByFieldName: { [x: string]: BackwardPolyRelationSpecType } = {};

    // const addField = (fieldName, description, type, resolve, spec, hint) => {
    //   // Field
    //   newFields = extend(
    //     newFields,
    //     {
    //       [fieldName]: fieldWithHooks(
    //         fieldName,
    //         {
    //           description,
    //           type,
    //         },
    //         {
    //           isPgConnectionFilterField: true,
    //         },
    //       ),
    //     },
    //     hint,
    //   );
    //   // Relation spec for use in resolver
    //   backwardRelationSpecByFieldName = extend(backwardRelationSpecByFieldName, {
    //     [fieldName]: spec,
    //   });
    //   // Resolver
    //   connectionFilterRegisterResolver(Self.name, fieldName, resolve);
    // };

    for (const spec of backwardRelationSpecs) {
      const { foreignTable, constraint, tablePrimaryKey, fieldName, isOneToMany } = spec;
      const foreignTableTypeName = inflection.tableType(foreignTable);
      const foreignTableFilterTypeName = inflection.filterType(foreignTableTypeName);
      const ForeignTableFilterType = connectionFilterType(
        newWithHooks,
        foreignTableFilterTypeName,
        foreignTable,
        foreignTableTypeName,
      );
      if (!ForeignTableFilterType) continue;

      if (isOneToMany) {
        if (!omit(foreignTable, 'many')) {
          const filterManyTypeName = inflection.filterManyPolyType(table, foreignTable);
          if (!connectionFilterTypesByTypeName[filterManyTypeName]) {
            connectionFilterTypesByTypeName[filterManyTypeName] = newWithHooks(
              GraphQLInputObjectType,
              {
                name: filterManyTypeName,
                description: `A filter to be used against many \`${foreignTableTypeName}\` object
                 through polymorphic types. All fields are combined with a logical ‘and.’`,
              },
              {
                foreignTable,
                isPgConnectionFilterManyPoly: true,
                backwardRelationSpec: spec,
              },
            );
          }
          const FilterManyType = connectionFilterTypesByTypeName[filterManyTypeName];
          newFields = addField(
            fieldName,
            `Filter by the object’s \`${fieldName}\` relation.`,
            FilterManyType,
            makeResolveMany(spec),
            spec,
            `Adding connection filter backward relation field from ${describePgEntity(
              table,
            )} to ${describePgEntity(foreignTable)}`,
            build,
            newFields,
            backwardRelationSpecByFieldName,
            context,
          );
        }
      } else {
        addField(
          fieldName,
          `Filter by the object’s \`${fieldName}\` relation.`,
          ForeignTableFilterType,
          resolveSingle,
          spec,
          `Adding connection filter backward relation field from ${describePgEntity(
            table,
          )} to ${describePgEntity(foreignTable)}`,
          build,
          newFields,
          backwardRelationSpecByFieldName,
          context,
        );
      }
    }

    function resolveSingle({ sourceAlias, fieldName, fieldValue, queryBuilder }) {
      if (fieldValue == null) return null;

      const {
        foreignTable,
        table,
        constraint,
        tablePrimaryKey,
      } = backwardRelationSpecByFieldName[fieldName];

      const foreignTableTypeName = inflection.tableType(foreignTable);
      const foreignTableAlias = sql.identifier(Symbol());

      const foreignTableFilterTypeName = inflection.filterType(foreignTableTypeName);

      const sqlSelectWhereKeysMatch = getSqlSelectWhereKeysMatch({
        sourceAlias, foreignTableAlias,
        foreignTable, table, constraint, tablePrimaryKey, sql, inflection,
      });

      const sqlFragment = connectionFilterResolve(
        fieldValue,
        foreignTableAlias,
        foreignTableFilterTypeName,
        queryBuilder,
      );
      return sqlFragment == null
        ? null
        : sql.query`exists(${sqlSelectWhereKeysMatch} and (${sqlFragment}))`;
    }

    function makeResolveMany(backwardRelationSpec) {
      return function resolveMany({ sourceAlias, fieldName, fieldValue, queryBuilder }) {
        if (fieldValue == null) return null;

        const { foreignTable } = backwardRelationSpecByFieldName[fieldName];

        const foreignTableFilterManyTypeName = inflection.filterManyPolyType(table, foreignTable);
        const sqlFragment = connectionFilterResolve(
          fieldValue,
          sourceAlias,
          foreignTableFilterManyTypeName,
          queryBuilder,
          null,
          null,
          null,
          { backwardRelationSpec },
        );
        return sqlFragment == null ? null : sqlFragment;
      };
    }
    return newFields;
  });

  builder.hook('GraphQLInputObjectType:fields', (fields, build, context) => {
    const {
      extend,
      newWithHooks,
      inflection,
      pgSql: sql,
      connectionFilterResolve,
      connectionFilterRegisterResolver,
      connectionFilterTypesByTypeName,
      connectionFilterType,
    } = build;
    const {
      fieldWithHooks,
      scope: { foreignTable, isPgConnectionFilterManyPoly, backwardRelationSpec },
      Self,
    } = context;

    if (!isPgConnectionFilterManyPoly || !foreignTable) return fields;

    connectionFilterTypesByTypeName[Self.name] = Self;

    const foreignTableTypeName = inflection.tableType(foreignTable);
    const foreignTableFilterTypeName = inflection.filterType(foreignTableTypeName);
    const FilterType = connectionFilterType(
      newWithHooks,
      foreignTableFilterTypeName,
      foreignTable,
      foreignTableTypeName,
    );

    const manyFields = {
      every: fieldWithHooks(
        'every',
        {
          description: `Every related \`${foreignTableTypeName}\` matches the filter criteria. All fields are combined with a logical ‘and.’`,
          type: FilterType,
        },
        {
          isPgConnectionFilterManyField: true,
        },
      ),
      some: fieldWithHooks(
        'some',
        {
          description: `Some related \`${foreignTableTypeName}\` matches the filter criteria. All fields are combined with a logical ‘and.’`,
          type: FilterType,
        },
        {
          isPgConnectionFilterManyField: true,
        },
      ),
      none: fieldWithHooks(
        'none',
        {
          description: `No related \`${foreignTableTypeName}\` matches the filter criteria. All fields are combined with a logical ‘and.’`,
          type: FilterType,
        },
        {
          isPgConnectionFilterManyField: true,
        },
      ),
    };

    const resolve = ({ sourceAlias, fieldName, fieldValue, queryBuilder }) => {
      if (fieldValue == null) return null;

      // foreignTable is the polymorphic table, like tags, notes,
      const {
        foreignTable, table, constraint, tablePrimaryKey,
      } = backwardRelationSpec as BackwardPolyRelationSpecType;
      const foreignTableAlias = sql.identifier(Symbol());
      const sqlSelectWhereKeysMatch = getSqlSelectWhereKeysMatch({
        sourceAlias, foreignTableAlias,
        foreignTable, table, constraint, tablePrimaryKey, sql, inflection,
      });
      const sqlFragment = connectionFilterResolve(
        fieldValue,
        foreignTableAlias,
        foreignTableFilterTypeName,
        queryBuilder,
      );
      if (sqlFragment == null) {
        return null;
      }
      if (fieldName === 'every') {
        return sql.query`not exists(${sqlSelectWhereKeysMatch} and not (${sqlFragment}))`;
      }
      if (fieldName === 'some') {
        return sql.query`exists(${sqlSelectWhereKeysMatch} and (${sqlFragment}))`;
      }
      if (fieldName === 'none') {
        return sql.query`not exists(${sqlSelectWhereKeysMatch} and (${sqlFragment}))`;
      }
      throw new Error(`Unknown field name: ${fieldName}`);
    };

    for (const fieldName of Object.keys(manyFields)) {
      connectionFilterRegisterResolver(Self.name, fieldName, resolve);
    }

    return extend(fields, manyFields);
  });
};
