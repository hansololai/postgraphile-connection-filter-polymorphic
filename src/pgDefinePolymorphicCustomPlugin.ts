import { SchemaBuilder, Options } from 'postgraphile';
export interface PgPolymorphicConstraint {
  name: string;
  from: string; // classId
  to: string[]; // due to limitation at the time, it is the ModelName array.
}
export type PgPolymorphicConstraintByName = PgPolymorphicConstraint[];

/**
 * @description This plugin add an array named 'pgPolymorphicClassAndTargetModels' in build,
 * If it already exist, it will append to it. It adds custom defined polymorphic constraints.
 * A polymorphic association via @smartComments on a xxx_type column should have 2 fields in tag
 * a isPolymorphic:true, (maybe can be deprecated), a polymorphicTo:[ModelNames].
 * @example create comment syntax
 * comment on column notes.noteable
 *  is E'@isPolymorphic\n@polymorphicTo Location\n@polymorphicTo Workflow'
 * @param builder The SchemaBuilder
 * @param options The option passed in. This Option is the same object allows access of custom
 * parameters they pass in when the call 'createPostGraphileSchema'
 * @author Han Lai
 */
export const definePolymorphicCustom = (builder: SchemaBuilder, options: Options) => {
  // First add an inflector for polymorphic backrelation type name
  builder.hook('inflection', inflection => ({
    ...inflection,
    forwardRelationByPolymorphic(table, polymorphicName: string) {
      return this.camelCase(`${this.singularize(table.name)}-as-${polymorphicName}`);
    },
  }));
  builder.hook('build', (build) => {
    const {
      pgSql: sql,
      pgIntrospectionResultsByKind: { class: pgClasses, attributeByClassIdAndNum },
      pgPolymorphicClassAndTargetModels = [],
    } = build;

    const { pgSchemas = [] } = options as any;
    const pgPolymorphicClassAndTargetModelsCustome: PgPolymorphicConstraintByName = pgClasses
      .filter(c => pgSchemas.includes(c.namespaceName) && c.classKind === 'r')
      .reduce((acc, curClass) => {
        const curClassAttributes: { [x: string]: any } = attributeByClassIdAndNum[curClass.id];
        // We do it in two steps, first find all xxx_type
        const allCurrentClassAttributes = Object.values(curClassAttributes);
        const typeAttributes = allCurrentClassAttributes.filter((attribute) => {
          // Must be a xxx_type attribute, and also tags need to have "isPolymorphic"
          return attribute.name.endsWith('_type')
            && (!!attribute.tags.isPolymorphic || attribute.type.type === 'e');
        });
        const polyConstraintsOfClass = typeAttributes.map((attribute) => {
          const { name, tags: { polymorphicTo = [] } } = attribute;
          const polymorphicKey = name.substring(0, name.length - 5);
          const newPolyConstraint: PgPolymorphicConstraint = {
            name: polymorphicKey,
            from: curClass.id,
            to: polymorphicTo,
          };
          return newPolyConstraint;
        });

        return [...acc, ...polyConstraintsOfClass];
      }, []);

    const pgPolymorphicClassAndTargetModelsCombined = [
      ...pgPolymorphicClassAndTargetModels, ...pgPolymorphicClassAndTargetModelsCustome,
    ];

    return build.extend(build, {
      pgPolymorphicClassAndTargetModels: pgPolymorphicClassAndTargetModelsCombined,
    });
  });
};
