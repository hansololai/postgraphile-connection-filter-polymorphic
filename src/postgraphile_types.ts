import { Build } from 'postgraphile';
import {
  PgAttribute,
  PgProc,
  PgClass,
  PgConstraint,
  PgExtension,
  PgType,
  PgNamespace,
} from 'graphile-build-pg';
import { PgIndex } from 'graphile-build-pg/node8plus/plugins/PgIntrospectionPlugin';

export type GraphilePgAttribute = PgAttribute;
export type GraphilePgProcedure = PgProc;

export interface GraphilePgConstraint extends Omit<PgConstraint, 'foreignClass' | 'class'> {
  foreignClass: GraphilePgClass;
  class: GraphilePgClass;
}
export interface GraphilePgClass extends Omit<PgClass, 'constraints' | 'attributes'> {
  constraints: GraphilePgConstraint[];
  attributes: PgAttribute[];
}
export interface GraphilePgIntrospection {
  __pgVersion: number;
  attribute: PgAttribute[];
  attributeByClassIdAndNum: { [classId: string]: { [num: string]: PgAttribute } };
  class: GraphilePgClass[];
  classById: { [x: string]: GraphilePgClass };
  constraint: GraphilePgConstraint[];
  extension: PgExtension[];
  extensionById: { [x: string]: PgExtension };
  index: PgIndex[];
  namespace: PgNamespace[];
  namespaceById: { [x: string]: PgNamespace };
  procedure: PgProc[];
  type: PgType[];
  typeById: { [x: string]: PgType };
}
export interface GraphileBuild extends Build {
  pgIntrospectionResultsByKind: GraphilePgIntrospection;
}
