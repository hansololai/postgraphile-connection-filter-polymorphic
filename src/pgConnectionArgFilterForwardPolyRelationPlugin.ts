import { SchemaBuilder, Options } from 'postgraphile';
import { PgPolymorphicConstraintByName, PgPolymorphicConstraint } from './pgDefinePolymorphicCustomPlugin';
import { addField } from './pgConnectionArgFilterBackwardPolyRelationPlugin';

export interface ForwardPolyRelationSpecType {
  table: any;
  foreignTable: any;
  fieldName: string;
  foreignPrimaryKey: any;
  constraint: PgPolymorphicConstraint;
}
export const addForwardPolyRelationFilter = (builder: SchemaBuilder) => {
  // builder.hook('inflection', (inflection) => ({
  //   ...inflection,
  //   filterForwardRelationExistsFieldName(relationFieldName) {
  //     return `${relationFieldName}Exists`;
  //   },
  // }));

  builder.hook('GraphQLInputObjectType:fields', (fields, build, context) => {
    const {
      describePgEntity,
      extend,
      newWithHooks,
      inflection,
      pgSql: sql,
      pgIntrospectionResultsByKind: introspectionResultsByKind,
      pgIntrospectionResultsByKind: { classById },
      connectionFilterResolve,
      connectionFilterRegisterResolver,
      connectionFilterTypesByTypeName,
      connectionFilterType,
      mapFieldToPgTable,
      pgPolymorphicClassAndTargetModels = [],
    } = build;
    const {
      fieldWithHooks,
      scope: { pgIntrospection: table, isPgConnectionFilter },
      Self,
    } = context;
    let newFields = fields;

    if (!isPgConnectionFilter || table.kind !== 'class') return fields;
    if (!mapFieldToPgTable) {
      throw new Error(
        'mapFieldToPgTable is not defined in build, \
        you might missed plugin to run before this plugin',
      );
    }
    // A function convert the modelName to table.id
    const reFormatPolymorphicConstraint = (cur: PgPolymorphicConstraint) => {
      const newTo = cur.to
        .map((targetModelName) => {
          const t = mapFieldToPgTable[targetModelName];
          if (!t) {
            return null;
          }
          return classById[t.id];
        })
        .filter((c) => {
          return c && c.classKind === 'r';
        })
        .map((r) => {
          return r.id;
        });
      return { ...cur, to: newTo };
    };

    connectionFilterTypesByTypeName[Self.name] = Self;

    // Iterate the pgPolymorphic constraints and find the ones that are relavent to this table
    const forwardPolyRelationSpecs: ForwardPolyRelationSpecType[]
      = (<PgPolymorphicConstraintByName>pgPolymorphicClassAndTargetModels)
        .filter(con => con.from === table.id)
        .reduce((acc, currentPoly) => {
          const cur = reFormatPolymorphicConstraint(currentPoly);
          // For each polymorphic, we collect the following, using Tag as example
          // Suppose Tag can be tagged on User, Post via taggable_id and taggable_type
          // 1. target table objects. e.g. User, Post
          // 2. fieldNames e.g. UserAsTaggable, PostAsTaggable
          // 3. constraint name. e.g. taggable
          // 4. foreignTableAttribute e.g. 'id'
          const toReturn: ForwardPolyRelationSpecType[] = cur.to.reduce(
            (memo, curForeignTable) => {
              const foreignTable = classById[curForeignTable];
              if (!foreignTable) return memo;
              const fieldName = inflection.forwardRelationByPolymorphic(foreignTable, cur.name);
              const foreignPrimaryConstraint = introspectionResultsByKind.constraint.find(
                attr => attr.classId === foreignTable.id && attr.type === 'p',
              );
              if (foreignPrimaryConstraint && foreignPrimaryConstraint.keyAttributes[0]) {
                memo.push({
                  table,
                  foreignTable,
                  fieldName,
                  foreignPrimaryKey: foreignPrimaryConstraint.keyAttributes[0],
                  constraint: currentPoly,
                });
              }

              return memo;
            },
            [],
          );
          return [...acc, ...toReturn];
        }, [] as ForwardPolyRelationSpecType[]);

    let forwardPolyRelationSpecByFieldName: { [x: string]: ForwardPolyRelationSpecType } = {};

    // const addField = (fieldName, description, type, resolve, spec, hint) => {
    //   // Field
    //   newFields = extend(
    //     newFields,
    //     {
    //       [fieldName]: fieldWithHooks(
    //         fieldName,
    //         {
    //           description,
    //           type,
    //         },
    //         {
    //           isPgConnectionFilterField: true,
    //         },
    //       ),
    //     },
    //     hint,
    //   );
    //   // Spec for use in resolver
    //   forwardPolyRelationSpecByFieldName = extend(forwardPolyRelationSpecByFieldName, {
    //     [fieldName]: spec,
    //   });
    //   // Resolver
    //   connectionFilterRegisterResolver(Self.name, fieldName, resolve);
    // };

    for (const spec of forwardPolyRelationSpecs) {
      const { foreignTable, fieldName } = spec;

      const foreignTableTypeName = inflection.tableType(foreignTable);
      const foreignTableFilterTypeName = inflection.filterType(foreignTableTypeName);
      const ForeignTableFilterType = connectionFilterType(
        newWithHooks,
        foreignTableFilterTypeName,
        foreignTable,
        foreignTableTypeName,
      );
      if (!ForeignTableFilterType) continue;

      newFields = addField(
        fieldName,
        `Filter by the objectâ€™s \`${fieldName}\` polymorphic relation.`,
        ForeignTableFilterType,
        resolve,
        spec,
        `Adding connection filter forward polymorphic relation field from ${describePgEntity(
          table,
        )} to ${describePgEntity(foreignTable)}`,
        build,
        newFields,
        forwardPolyRelationSpecByFieldName,
        Self
      );
    }

    function resolve({ sourceAlias, fieldName, fieldValue, queryBuilder }) {
      if (fieldValue == null) return null;

      const {
        foreignTable,
        foreignPrimaryKey,
        constraint,
      } = forwardPolyRelationSpecByFieldName[fieldName];

      const foreignTableTypeName = inflection.tableType(foreignTable);
      const foreignTableAlias = sql.identifier(Symbol());
      const sourceTableId = `${constraint.name}_id`;
      const sourceTableType = `${constraint.name}_type`;

      const sqlIdentifier = sql.identifier(foreignTable.namespace.name, foreignTable.name);
      // sql match query
      // sql string "(table_alias).xxx_type = 'User' and (table alias).xxx_id = (users alias).id"
      const sqlKeysMatch = sql.query`(${sql.fragment`${sourceAlias}.${sql.identifier(
        sourceTableId,
      )} = ${foreignTableAlias}.${sql.identifier(foreignPrimaryKey.name)}`}) and (
        ${sql.fragment`${sourceAlias}.${sql.identifier(
        sourceTableType,
      )} = ${sql.value(foreignTableTypeName)}`})`;

      const foreignTableFilterTypeName = inflection.filterType(foreignTableTypeName);

      const sqlFragment = connectionFilterResolve(
        fieldValue,
        foreignTableAlias,
        foreignTableFilterTypeName,
        queryBuilder,
      );

      return sqlFragment == null
        ? null
        : sql.query`\
      exists(
        select 1 from ${sqlIdentifier} as ${foreignTableAlias}
        where ${sqlKeysMatch} and
          (${sqlFragment})
      )`;
    }

    return newFields;
  });
};
