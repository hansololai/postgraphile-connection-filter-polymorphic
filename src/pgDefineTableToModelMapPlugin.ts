import { SchemaBuilder, Options } from 'postgraphile';
import { GraphilePgAttribute, GraphileBuild, GraphilePgProcedure } from './postgraphile_types';
interface SimplePgTableIntrospect {
  name: string;
  id: string;
  attributesMap: AttributesMap;
}
interface AttributesMap {
  [x: string]: GraphilePgAttribute | GraphilePgProcedure;
}
export type FieldToDBMap = {
  [x: string]: SimplePgTableIntrospect;
};

export const addModelTableMappingPlugin = (builder: SchemaBuilder, options: Options) => {
  const { pgSchemas = [] } = options as any;
  builder.hook('build', (build) => {
    const {
      pgIntrospectionResultsByKind: { procedure, class: pgClasses },
      inflection: { upperCamelCase, singularize, camelCase },
    } = build as GraphileBuild;

    const fieldToDBMap: FieldToDBMap = pgClasses.reduce((acc: FieldToDBMap, cur) => {
      // Only build the map for the included schema.
      // Also this mapping behave unexpectedly if there are tables with same name in different
      // Schema used in the postgraphile
      if (
        !pgSchemas.includes(cur.namespaceName) ||
        cur.namespaceName === 'pg_catalog' ||
        cur.namespaceName === 'information_schema'
      ) {
        // skipt it
        return acc;
      }
      const procedureAttriutesMap = procedure
        .filter(p => p.name.startsWith(`${cur.name}_`))
        .reduce((a: AttributesMap, c) => {
          // Should probably use inflection
          const k = singularize(camelCase(c.name.replace(`${cur.name}_`, '')));
          a[k] = c;
          return a;
        }, {});
      // The Model Name points to this pg object
      const curTable: SimplePgTableIntrospect = {
        name: cur.name,
        id: cur.id,
        attributesMap: cur.attributes.reduce((allAtt: AttributesMap, curA) => {
          allAtt[singularize(camelCase(curA.name))] = curA;
          return allAtt;
        }, procedureAttriutesMap),
      };
      acc[singularize(upperCamelCase(cur.name))] = curTable;
      return acc;
    }, {});
    return build.extend(build, { mapFieldToPgTable: fieldToDBMap });
  });
};
