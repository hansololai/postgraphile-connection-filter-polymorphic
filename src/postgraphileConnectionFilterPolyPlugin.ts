import { addBackwardPolyRelationFilter } from './pgConnectionArgFilterBackwardPolyRelationPlugin';
import { addForwardPolyRelationFilter } from './pgConnectionArgFilterForwardPolyRelationPlugin';
import { definePolymorphicCustom } from './pgDefinePolymorphicCustomPlugin';
import { addModelTableMappingPlugin } from './pgDefineTableToModelMapPlugin';
import {
  SchemaBuilder, Options, makePluginByCombiningPlugins,
} from 'postgraphile';
const PostGraphileConnectionFilterPolyCorePlugin = makePluginByCombiningPlugins(
  addModelTableMappingPlugin,
  definePolymorphicCustom,
  addForwardPolyRelationFilter,
  addBackwardPolyRelationFilter,
);

export const PostGraphileConnectionFilterPolyPlugin = (
  builder: SchemaBuilder, options: Options,
) => {
  builder.hook('build', (build) => {
    const pkg = require('../package.json');

    // Check dependencies
    if (!build.versions) {
      throw new Error(
        `Plugin ${pkg.name}@${
        pkg.version
        } requires graphile-build@^4.1.0 in order to check dependencies (current version: ${
        build.graphileBuildVersion
        })`,
      );
    }
    const depends = (name: string, range: string) => {
      if (!build.hasVersion(name, range)) {
        throw new Error(
          `Plugin ${pkg.name}@${pkg.version} requires ${name}@${range} (${
          build.versions[name]
            ? `current version: ${build.versions[name]}`
            : 'not found'
          })`,
        );
      }
    };
    depends('postgraphile-plugin-connection-filter', '^1.0.0');

    // Register this plugin
    build.versions = build.extend(build.versions, { [pkg.name]: pkg.version });

    return build;
  });

  PostGraphileConnectionFilterPolyCorePlugin(builder, options);
};
