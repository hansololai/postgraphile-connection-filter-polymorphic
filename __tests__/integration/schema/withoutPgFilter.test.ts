import { PostGraphileConnectionFilterPolyPlugin } from '../../../src/';
import { withPgClient } from '../../helpers';
import { createPostGraphileSchema } from 'postgraphile';
test('prints a schema with the filter plugin', async () => {
  await expect(
    withPgClient(async (client) => {
      await createPostGraphileSchema(client, ['p'], {
        appendPlugins: [PostGraphileConnectionFilterPolyPlugin],
        disableDefaultMutations: true,
      });
    }),
    // this should give error because the PgConnectionFilter plugin is not appended
  ).rejects.toThrowError();
});
