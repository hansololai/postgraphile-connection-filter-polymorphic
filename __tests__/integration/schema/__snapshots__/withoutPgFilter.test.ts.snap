// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints a schema with the filter plugin 1`] = `"Plugin postgraphile-plugin-connection-filter-polymorphic@1.1.12 requires postgraphile-plugin-connection-filter@^1.0.0 (not found)"`;
